File: examples/nonlinearity/fewbit.py
# Copyright 2024 Daniel Bershatsky
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#     http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.

from functools import partial

import jax
import jax.numpy as jnp
import numpy as np
from jax import Array
from jax.core import ShapedArray
from jax.extend.core import Primitive
from jax.interpreters import mlir, xla
from jax.lib import xla_client
from jaxlib.hlo_helpers import custom_call

import _fewbit

__all__ = ('gelu', 'gelu_fwd', 'grandient_quantized_bwd')

for name, fn in _fewbit.registrations().items():
    xla_client.register_custom_call_target(name, fn)

BOUNDARIES = jnp.array([
    -1.00000000e+02, -2.41658117e+00, -7.10008030e-01, -3.25840561e-01,
    1.06942188e-04, 3.26057167e-01, 7.10240866e-01, 2.41447889e+00,
    1.00000000e+02
])[1:-1]

LEVELS = jnp.array([
    -1.93991237e-04, -8.82791291e-02, 1.25683834e-01, 3.72314415e-01,
    6.27851368e-01, 8.74450518e-01, 1.08834805e+00, 1.00019494e+00
])


grandient_quantized_bwd_p = Primitive('fewbit::grandient_quantized_bwd')
grandient_quantized_bwd_p.multiple_results = False


def grandient_quantized_bwd(xs: Array, levels: Array,
                            cotangents: Array) -> Array:
    """Compute VJP of gradient-quantized activation function."""
    return grandient_quantized_bwd_p.bind(xs, levels, cotangents)


@grandient_quantized_bwd_p.def_impl
def grandient_quantized_bwd_impl(xs: Array, levels: Array,
                                 cotangents: Array) -> Array:
    return xla.apply_primitive(grandient_quantized_bwd_p, xs, levels,
                               cotangents)


@grandient_quantized_bwd_p.def_abstract_eval
def grandient_quantized_bwd_abstract_eval(xs: Array, levels: Array,
                                          cotangents: Array) -> ShapedArray:
    return ShapedArray(cotangents.shape, cotangents.dtype)


@partial(mlir.register_lowering, grandient_quantized_bwd_p)
def grandient_quantized_bwd_lowering(ctx: mlir.LoweringRuleContext, xs, levels,
                                     cotangents):
    # Describe input and output buffers.
    bits = np.int64(levels.type.shape[0].bit_length() - 1)
    dtype = mlir.ir.RankedTensorType(cotangents.type)
    size = np.prod(dtype.shape).astype(np.int64)
    layout = tuple(range(len(dtype.shape) - 1, -1, -1))

    op = custom_call(
        'cpu_gradient_quantized_bwd_f32',
        operands=[
            mlir.ir_constant(bits), xs, levels,
            mlir.ir_constant(size), cotangents
        ],
        operand_layouts=[(), (0, ), (0, ), (), layout],
        result_types=[dtype],
        result_layouts=[layout],
    )
    return op.results


gelu_fwd_p = Primitive('fewbit::gelu_fwd')
gelu_fwd_p.multiple_results = True
gelu_fwd_p.def_impl(partial(xla.apply_primitive, gelu_fwd_p))


def gelu_fwd(xs: Array, boundaries: Array) -> Array:
    """Compute VJP-forward rule of gradient-quantized GELU."""
    return gelu_fwd_p.bind(xs, boundaries)


@gelu_fwd_p.def_abstract_eval
def gelu_fwd_abstract_eval(xs: Array, boundaries: Array):
    bits = boundaries.size.bit_length()
    total_bytes = (xs.size * bits) // 8
    if (xs.size * bits) % 8 > 0:
        total_bytes += 1
    return ShapedArray(xs.shape, xs.dtype), ShapedArray(total_bytes, jnp.uint8)


@partial(mlir.register_lowering, gelu_fwd_p)
def gelu_fwd_lowering(ctx: mlir.LoweringRuleContext, xs: Array, bs: Array):
    # Describe input buffers.
    dtype = mlir.ir.RankedTensorType(xs.type)
    size = np.prod(dtype.shape).astype(np.int64)
    layout = tuple(range(len(dtype.shape) - 1, -1, -1))

    # Describe output buffers.
    bits = np.int64(bs.type.shape[0].bit_length())
    length = (size * bits) // 8
    if (size * bits) % 8 > 0:
        length += 1
    elem_dtype = mlir.ir.IntegerType.get_unsigned(8)
    accum_dtype = mlir.ir.RankedTensorType.get([length], elem_dtype)

    op = custom_call(
        'cpu_gelu_fwd_f32',
        operands=[mlir.ir_constant(size), xs,
                  mlir.ir_constant(bits), bs],
        operand_layouts=[(), layout, (), (0, )],
        result_types=[dtype, accum_dtype],
        result_layouts=[layout, (0, )],
    )
    return op.results


gelu_p = Primitive('fewbit_gelu')
gelu_p.multiple_results = False
gelu_p.def_impl(partial(xla.apply_primitive, gelu_p))


@jax.custom_vjp
def gelu(xs: Array, boundaries: Array, levels: Array) -> Array:
    """Calculate activations of gradient-quantized GELU.

    Args:
      xs: Inputs.
      boundaries: Gradient quantization boundaries.
      levels: Gradient quantization levels.

    Returns:
      GELU activations.
    """
    return gelu_p.bind(xs, boundaries, levels)


@gelu_p.def_abstract_eval
def gelu_abstract_eval(xs: Array, boundaries: Array,
                       levels: Array) -> ShapedArray:
    return ShapedArray(xs.shape, xs.dtype)


@partial(mlir.register_lowering, gelu_p)
def gelu_lowering(ctx: mlir.LoweringRuleContext, xs: Array, bs: Array,
                  ls: Array):
    # Describe input buffers.
    dtype = mlir.ir.RankedTensorType(xs.type)
    size = np.prod(dtype.shape).astype(np.int64)
    layout = tuple(range(len(dtype.shape) - 1, -1, -1))

    op = custom_call(
        'cpu_gelu_f32',
        operands=[mlir.ir_constant(size), xs],
        operand_layouts=[(), layout],
        result_types=[dtype],
        result_layouts=[layout],
    )
    return op.results


def gelu_vjp_fwd(inputs: Array, boundaries: Array, levels: Array):
    outputs, primals = gelu_fwd(inputs, boundaries)
    return outputs, (primals, levels)


def gelu_vjp_bwd(residue: tuple[Array, Array], cotangents: Array):
    primals, levels = residue
    results = grandient_quantized_bwd(primals, levels, cotangents)
    return (results, None, None)


gelu.defvjp(gelu_vjp_fwd, gelu_vjp_bwd)

File: examples/nonlinearity/fewbit.cc
/**
 * Copyright 2024 Daniel Bershatsky
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

#include <cmath>
#include <span>

#include <nanobind/nanobind.h>

#include "kernel_nanobind_helpers.h"

namespace nb = nanobind;

namespace fewbit {

constexpr long double kSqrt2 = 1.4142135623730950488;

template <typename T> constexpr T Gelu(T value) {
    constexpr T sqrt2 = kSqrt2;
    return 0.5 * value * (1 + std::erf(value / sqrt2));
}

template <typename T> void Gelu(void *out, void const **ins) {
    auto size = *reinterpret_cast<int64_t const *>(ins[0]);
    auto xs = reinterpret_cast<T const *>(ins[1]);
    auto ys = reinterpret_cast<T *>(out);

    for (int64_t ix = 0; ix != size; ++ix) {
        ys[ix] = Gelu(xs[ix]);
    }
}

template <typename T>
void GeluFwd(std::span<T const> input, T *output, uint8_t *primals,
             std::span<T const> boundaries) {
    int bit_width = std::popcount(boundaries.size());
    int max_width = 8 * sizeof(uint8_t);
    int threshold = max_width - bit_width;
    int shift = 0;
    for (auto const in : input) {
        // Compute activation function output.
        *output = Gelu(in);
        ++output;

        // Quantize input.
        auto pos = std::lower_bound(boundaries.begin(), boundaries.end(), in);
        auto ix = std::distance(boundaries.begin(), pos);

        // Compress quantization index to bit stream.
        if (shift <= threshold) {
            *primals |= static_cast<uint8_t>(ix << shift);
            shift = (shift + bit_width) % max_width;
        } else {
            auto lo = (ix << shift) & 0xff;
            auto hi = ix >> (max_width - shift);
            shift = shift + bit_width - max_width;
            primals[0] |= static_cast<uint8_t>(lo);
            primals[1] = static_cast<uint8_t>(hi);
            ++primals;
        }
    }
}

template <typename T> void GeluFwd(void *out, void const **ins) {
    auto size = *reinterpret_cast<uint64_t const *>(ins[0]);
    auto xs = reinterpret_cast<T const *>(ins[1]);
    auto bits = *reinterpret_cast<uint64_t const *>(ins[2]);
    auto bs = reinterpret_cast<T const *>(ins[3]); // Boundaries.

    auto outs = reinterpret_cast<void **>(out);
    auto ys = reinterpret_cast<T *>(outs[0]);
    auto ps = reinterpret_cast<uint8_t *>(outs[1]);

    GeluFwd<T>({xs, size}, ys, ps, {bs, (1u << bits) - 1});
}

template <typename T>
void GradientQuantizedBwd(std::span<uint8_t const> primals,
                          std::span<T const> out_cotangets, T *in_cotangents,
                          std::span<T const> levels) {
    int bit_width = std::popcount(levels.size() - 1);
    int max_width = 8 * sizeof(uint8_t);
    int shift = 0;
    auto mask = (1u << bit_width) - 1;
    auto primal = primals.begin();
    for (auto out_cotangent : out_cotangets) {
        // Inflate compressed quantization level indices.
        uint64_t index = (*primal >> shift) & mask;
        shift = shift + bit_width;
        if (shift > max_width) {
            ++primal;
            shift -= max_width;
            index |= (*primal << (bit_width - shift)) & mask;
        }

        // Scale input cotangents with quantization level.
        *in_cotangents = levels[index] * out_cotangent;
        ++in_cotangents;
    }
}

template <typename T> void GradientQuantizedBwd(void *out, void const **ins) {
    auto bits = *reinterpret_cast<uint64_t const *>(ins[0]);
    auto ps = reinterpret_cast<uint8_t const *>(ins[1]); // Primals
    auto ls = reinterpret_cast<T const *>(ins[2]);       // Levels
    auto size = *reinterpret_cast<uint64_t const *>(ins[3]);
    auto xs = reinterpret_cast<T const *>(ins[4]);

    auto ys = reinterpret_cast<T *>(out);

    constexpr auto max_width = 8 * sizeof(uint8_t);
    auto length = (bits * size) / max_width;
    if ((bits * size) % max_width > 0) {
        ++length;
    }

    GradientQuantizedBwd<T>({ps, length}, {xs, size}, ys, {ls, 1u << bits});
}

nb::capsule EncapsulateFunction(void (*fn)(void *, void const **)) {
    return jax::EncapsulateFunction(fn);
}

nb::dict Registrations() {
    nb::dict dict;
    dict["cpu_gelu_f32"] = EncapsulateFunction(Gelu<float>);
    dict["cpu_gelu_f64"] = EncapsulateFunction(Gelu<double>);
    dict["cpu_gelu_fwd_f32"] = EncapsulateFunction(GeluFwd<float>);
    dict["cpu_gelu_fwd_f64"] = EncapsulateFunction(GeluFwd<double>);
    dict["cpu_gradient_quantized_bwd_f32"] =
        EncapsulateFunction(GradientQuantizedBwd<float>);
    dict["cpu_gradient_quantized_bwd_f64"] =
        EncapsulateFunction(GradientQuantizedBwd<double>);
    return dict;
}

NB_MODULE(_fewbit, m) {
    m.def("registrations", &Registrations);
}

} // namespace fewbit
